lib/Dialect/FIRRTL/Transforms/WireDFT.cpp

```diff
@@ -135,49 +139,101 @@ void WireDFTPass::runOnOperation() {
+      if (anno.isClass(dftTestModeEnableAnnoClass)) {
+        // If we have already found a DFT enable, emit an error.
         if (enableSignal) {
           auto diag =
-              op->emitError("more than one thing marked as a DFT enable");
+              module->emitError("more than one thing marked as a DFT enable");
           diag.attachNote(enableSignal.getLoc()) << "first thing defined here";
           error = true;
           return false;
         }
         // Grab the enable value and remove the annotation.
-        enableSignal = getValueByFieldID(
-            ImplicitLocOpBuilder::atBlockEnd(op->getLoc(), op->getBlock()),
-            op->getResult(0), anno.getFieldID());
+        enableSignal =
+            getValueByFieldID(ImplicitLocOpBuilder::atBlockBegin(
+                                  module->getLoc(), module.getBodyBlock()),
+                              module.getArgument(i), anno.getFieldID());
         enableModule = module;
         return true;
+      }
+
+      // Check for clock div bypass signal.
+      if (anno.isClass(dftClockDividerBypassAnnoClass)) {
+        // If we have already found a DFT clock div bypass signal, emit an
+        // error.
+        if (clockDivBypassSignal) {
+          module->emitError(
+                    "more than one thing marked as a DFT clock div bypass")
+                  .attachNote(clockDivBypassSignal.getLoc())
+              << "first thing defined here";
+          error = true;
+          return false;
+        }
+        clockDivBypassSignal =
+            getValueByFieldID(ImplicitLocOpBuilder::atBlockBegin(
+                                  module->getLoc(), module.getBodyBlock()),
+                              module.getArgument(i), anno.getFieldID());
+        clockDivBypassModule = module;
+        return true;
+      }
+
+      return false;
+    });
+    if (error)
+      return signalPassFailure();
+
+    // Walk the module body looking for any operation marked as the DFT enable.
+    auto walkResult = module->walk([&](Operation *op) {
+      AnnotationSet::removeAnnotations(op, [&](Annotation anno) {
+        // If we have already encountered an error, continue.
+        if (error)
+          return false;
+
+        // Check for enable signal annotation.
+        if (anno.isClass(dftTestModeEnableAnnoClass)) {
+          if (enableSignal) {
+            auto diag =
+                op->emitError("more than one thing marked as a DFT enable");
+            diag.attachNote(enableSignal.getLoc())
+                << "first thing defined here";
+            error = true;
+            return false;
+          }
+          // Grab the enable value and remove the annotation.
+          enableSignal = getValueByFieldID(
+              ImplicitLocOpBuilder::atBlockEnd(op->getLoc(), op->getBlock()),
+              op->getResult(0), anno.getFieldID());
+          enableModule = module;
+          return true;
+        }
```

Comment:
Would be nice if there was some code sharing for this block.  I think there are some better helpers so maybe we can have 1 helper module ports and inner operations?  Also emitting the error on the block argument instead of the module will be better.
```c++
auto handleAnnotation = [&](Value value, Annotation anno, StringRef annoName, Value &signal, Operation *signalModule, ) {
        if (anno.isClass(annoName)) {
          if (signal) {
            auto diag = emitError(value.getLoc(), "more than one thing marked as ") << annoName;
            diag.attachNote(signal.getLoc()) << "first thing defined here";
            error = true;
            return false;
          }
          // Grab the enable value and remove the annotation.
          signal = getValueByFieldID(
              ImplicitLocOpBuilder::atBlockEnd(value.getLoc(), value.getBlock()),
              value, anno.getFieldID());
          signalModule = module;
          return true;
        }
```

Reply:
Awesome, thank you (both)! This is much nicer, will update shortly.